Mathlib/Analysis/Calculus/Gradient/Basic.lean

```diff
@@ -0,0 +1,370 @@
+`f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` where `x'` converges to `x`. -/
+def Gradient (f : F → 𝕜) (x : F) : F := (toDual 𝕜 F).symm (fderiv 𝕜 f x)
+
+local notation "∇" => Gradient
+
+variable {s : Set F} {L : Filter F}
+
+theorem HasGradientWithinAt_iff_HasFDerivWithinAt {s : Set F} :
+  HasGradientWithinAt f f' s x ↔ HasFDerivWithinAt f (toDual 𝕜 F f') s x := by
+  constructor <;> (intro h; exact h)
+
+theorem HasGradientWithinAt.hasFDerivWithinAt {s : Set F} (h : HasGradientWithinAt f f' s x) :
+  HasFDerivWithinAt f (toDual 𝕜 F f') s x := HasGradientWithinAt_iff_HasFDerivWithinAt.mp h
+
+theorem HasFDerivWithinAt.hasGradientWithinAt {s : Set F}
+  (h : HasFDerivWithinAt f (toDual 𝕜 F f') s x) :
+  HasGradientWithinAt f f' s x := HasGradientWithinAt_iff_HasFDerivWithinAt.mpr h
+
+theorem HasGradientAt_iff_HasFDerivAt :
+  HasGradientAt f f' x ↔ HasFDerivAt f (toDual 𝕜 F f') x := by
+  constructor <;> (intro h; exact h)
+
+theorem HasGradientAt.hasFDerivAt (h : HasGradientAt f f' x) :
+  HasFDerivAt f (toDual 𝕜 F f') x := HasGradientAt_iff_HasFDerivAt.mp h
+
+theorem HasFDerivAt.hasGradientAt (h : HasFDerivAt f (toDual 𝕜 F f') x) :
+  HasGradientAt f f' x := HasGradientAt_iff_HasFDerivAt.mpr h
+
+theorem Gradient_zero_of_not_differentiableAt (h : ¬DifferentiableAt 𝕜 f x) : ∇ f x = 0 := by
+  have : (toDual 𝕜 F).symm 0 = 0 := by simp only [map_zero]
+  rw [← fderiv_zero_of_not_differentiableAt h] at this
+  exact this
+
+local notation "⟪" x ", " y "⟫" => @inner 𝕜 _ _ x y
+
+theorem toDual_eq_inner (y x : F) : ((toDual 𝕜 F) y) x = ⟪y, x⟫ := rfl
+
+theorem toDual_of_toDual_frechet (frechet : F →L[𝕜] 𝕜) :
+  frechet = (toDual 𝕜 F) ((toDual 𝕜 F).symm frechet) := by
+  apply ContinuousLinearMap.ext_iff.mpr
+  intro x
+  have : ((toDual 𝕜 F) ((toDual 𝕜 F).symm frechet)) x = ⟪(toDual 𝕜 F).symm frechet, x⟫ := rfl
+  rw [this]
+  apply toDual_symm_apply.symm
+
+theorem toDual_of_toDual_gradient (gradient : F) :
+  gradient = (toDual 𝕜 F).symm ((toDual 𝕜 F) gradient) :=
+   Eq.symm (LinearIsometryEquiv.symm_apply_apply (toDual 𝕜 F) gradient)
+
+section GradientUniqueness
+
+variable {gradf gradg : F}
+
+theorem HasGradientAt.unique (hf : HasGradientAt f gradf x) (hg : HasGradientAt f gradg x) :
+  gradf = gradg := (toDual 𝕜 F).injective ((hf.hasFDerivAt).unique hg.hasFDerivAt)
+
+end GradientUniqueness
+
+theorem HasGradientWithinAt_iff_HasFDerivWithinAt' {s : Set F} {frechet : F →L[𝕜] 𝕜} :
+  HasGradientWithinAt f ((toDual 𝕜 F).symm frechet) s x ↔ HasFDerivWithinAt f frechet s x := by
+  constructor
+  · intro h
+    rw [toDual_of_toDual_frechet frechet]
+    exact h.hasFDerivWithinAt
+  · intro h
+    rw [toDual_of_toDual_frechet frechet] at h
+    exact h.hasGradientWithinAt
+
+theorem HasGradientAt.hasFDerivAt' {frechet : F →L[𝕜] 𝕜}
+  (h : HasGradientAt f ((toDual 𝕜 F).symm frechet) x) :
+    HasFDerivAt f frechet x := by
+      rw [toDual_of_toDual_frechet frechet]
+      exact h.hasFDerivAt
```

Comment:
```suggestion
    (h : HasGradientAt f ((toDual 𝕜 F).symm frechet) x) :
    HasFDerivAt f frechet x := by
  rw [toDual_of_toDual_frechet frechet]
  exact h.hasFDerivAt
```
Your indentation is wrong. The rule is: subsequent lines in a theorem statement indented by 4 spaces, proofs indented by 2 spaces (plus additional spaces in subproofs). Same thing in the rest of the file.

Reply:
Thanks, we will revise the second file in this rule!